src/software/ai/hl/stp/play/free_kick/free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,317 @@
+        {
+            chip_target = friendly.position();
+        }
+    }
+
+    chip_tactic->updateControlParams(event.common.world_ptr->ball().position(),
+                                     chip_target);
+    tactics_to_run[0].emplace_back(chip_tactic);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+void FreeKickPlayFSM::lookForPass(const FreeKickPlayFSM::Update &event)
+{
+    PriorityTacticVector tactics_to_run = {{}};
+
+    // Keep the kicker aligned to the ball
+    updateAlignToBallTactic(event.common.world_ptr);
+    tactics_to_run[0].emplace_back(align_to_ball_tactic);
+
+    // Already assigned one tactic to align to ball.
+    // Assign up to 2 receivers and the remaining tactics are defenders
+    setReceiverAndDefenderTactics(tactics_to_run, event, 2, 1);
+
+    // Find the current best pass
+    // Avoid passes to the goalie and the passing robot
+    std::vector<RobotId> robots_to_ignore = {};
+    auto friendly_goalie_id_opt = event.common.world_ptr->friendlyTeam().getGoalieId();
+    if (friendly_goalie_id_opt.has_value())
+    {
+        robots_to_ignore.push_back(friendly_goalie_id_opt.value());
+    }
+    auto robot_with_ball_opt = event.common.world_ptr->friendlyTeam().getNearestRobot(
+        event.common.world_ptr->ball().position());
+    if (robot_with_ball_opt.has_value())
+    {
+        robots_to_ignore.push_back(robot_with_ball_opt.value().id());
+    }
+    best_pass_and_score_so_far =
+        pass_generator.getBestPass(*event.common.world_ptr, robots_to_ignore);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+bool FreeKickPlayFSM::passFound(const Update &event)
+{
+    double time_since_pass_optimization_start_seconds =
+        (event.common.world_ptr->getMostRecentTimestamp() - pass_optimization_start_time)
+            .toSeconds();
+
+    double abs_min_pass_score =
+        ai_config.shoot_or_pass_play_config().abs_min_pass_score();
+    double min_perfect_pass_score =
+        ai_config.shoot_or_pass_play_config().min_perfect_pass_score();
+    double pass_score_ramp_down_duration =
+        ai_config.free_kick_play_config().max_time_commit_to_pass_seconds();
+
+    // To get the best pass possible we start by aiming for a perfect one and then
+    // decrease the minimum score over time
+    double min_score =
+        min_perfect_pass_score - std::min(time_since_pass_optimization_start_seconds /
+                                              pass_score_ramp_down_duration,
+                                          min_perfect_pass_score - abs_min_pass_score);
+
+    return best_pass_and_score_so_far.rating > min_score;
+}
+
+bool FreeKickPlayFSM::shouldAbortPass(const Update &event)
+{
+    // Check if ball has already been passed
+    if (distance(event.common.world_ptr->ball().position(),
+                 best_pass_and_score_so_far.pass.passerPoint()) >
+        BALL_IN_PLAY_DISTANCE_THRESHOLD_METERS)
+    {
+        return false;
+    }
```

Comment:
in the future, we should have a TODO to consolidate this type of logic into a `isPassed` function under `//software/ai/evaluation`